Sumo-Logic-Tools/Event_Based_S3_Automation/script.py

```diff
@@ -0,0 +1,265 @@
+
+	def makeGetRequest(self, url):
+		return requests.get(url = self.base_url + str(url), auth = HTTPBasicAuth(self.access_id, self.access_key))
+
+	def getSourcesLinks(self, listOfCollectors):
+		listOfSourcesLink = list()
+		try:
+			for collector in listOfCollectors['collectors']:
+				links = collector['links']
+				listOfSourcesLink.extend(self.extractSourcesLink(collector['links']))
+		except KeyError as e:
+			try:
+				collector = listOfCollectors['collector']
+				links = collector['links']
+				listOfSourcesLink.extend(self.extractSourcesLink(collector['links']))
+			except KeyError as e:
+				print("The API did not return any collectors!")
+		return listOfSourcesLink
+
+	def extractSourcesLink(self, links):
+		listOfSourcesLink = list()
+		for link in links:
+			if (link['rel'] == 'sources'):
+				listOfSourcesLink.append(link['href'])
+
+		return listOfSourcesLink
+
+	def getPollingSources(self, listOfSourcesLink):
+		listOfPollingSources = list()
+		for sourceLink in listOfSourcesLink:
+			allSources = self.makeGetRequest(sourceLink).json()
+			for source in allSources['sources']:
+				if (self.isPollingSource(source) == True):
+					listOfPollingSources.append(source)
+
+		return listOfPollingSources
+
+	def isPollingSource(self, source):
+		# get returns None if contentType is not available. The source cannot be polling if it does not have a content type
+		contentType = source.get('contentType')
+		return (contentType in self.pollingTypeSources)
+
+	def populateMapOfBucketNameToEndpoint(self, listOfPollingSources):
+		for source in listOfPollingSources:
+			bucketName = self.getBucketName(source)
+			endpoint = self.getEndpoint(source)
+			if (bucketName is not None and endpoint is not None):
+				self.mapOfBucketNameToEndpoint[bucketName].append(endpoint)
+
+	def getBucketName(self, source):
+		try:
+			thirdPartyRef = source['thirdPartyRef']
+			resources = thirdPartyRef['resources']
+			path = resources[0]['path']
+			return path['bucketName']
+		except:
+			return None
+
+	def getEndpoint(self, source):
+		return source.get('url')
+
+
+'''
+Helps in setting up event based S3 using the appropriate AWS APIs.
+It does the following tasks:
+1. Using the map of bucket name to endpoint generated above, it creates a topic per bucket. (So a topic for every key in that map)
+2. Configure event notification on the bucket to send the object creation notification to the corresponding topic created in step 1.
+3. Creates a subscription per endpoint for the correspodning topic to allow sending the notifications to Sumo Logic.
+'''
+class AwsSetupHelper:
+
+	mapOfBucketNameToTopicArn = {}
+
+	# AWS Clients
+	snsClient = boto3.client('sns')
+	s3Client = boto3.client('s3')
+	stsClient = boto3.client('sts')
+
+	def __init__(self, mapOfBucketNameToEndpoint):
+		self.mapOfBucketNameToEndpoint = mapOfBucketNameToEndpoint
+
+	def setupEventNotifications(self):
+		self.createTopicPerBucket()
+		self.createSubscriptionsPerTopic()
```

Comment:
In `createTopicPerBucket` and `createSubscriptionsPerTopic` I would recommend printing some status that confirms each topic and subsciption as it is being created. This give the user confidence that the script is working, and also provides a useful reference that includes all the names of the various entities that have been created.